### PR TITLE
WIP: Windows unit tests - move to junit

### PR DESCRIPTION
**Pull request by vikramhh**
Part of work required for https://github.com/moby/moby/pull/39675 on Windows.
https://github.com/moby/moby/pull/39638 covers the corresponding work on Linux.

Signed-off-by: Vikram bir Singh <vikrambir.singh@docker.com>

**- What I did**
Added functionality to Unit test execution code to build and use gotestsum.exe for test execution 

**- How I did it**
Modified the Unit test execution code in hack\make.ps1

**- How to verify it**
Execute unit tests and verify that test output is present in the bundles directory under root.

**- Description for the changelog**

Build and use gotestsum.exe for unit test execution\

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: hack/make.ps1 [hunk 1]
```diff
@@ -307,6 +309,44 @@ Function Validate-GoFormat($headCommit, $upstreamCommit) {
     }
 }
 
+Function Build-GoTestSum() {
+    if (Test-Path "$GOTESTSUM_LOCATION\gotestsum.exe") {
+        Write-Host "INFO: gotestsum already avaiable - skipping building it...."
```
Comment (thaJeztah) [inline hack/make.ps1]:
```suggestion
        Write-Host "INFO: gotestsum already available - skipping building it...."
```

File: hack/make.ps1 [hunk 2]
```diff
@@ -307,6 +309,44 @@ Function Validate-GoFormat($headCommit, $upstreamCommit) {
     }
 }
 
+Function Build-GoTestSum() {
```
Comment (thaJeztah) [inline hack/make.ps1]: Should we build this as part of the Dockerfile, and then copy the binary out?

File: hack/make.ps1 (hunk 1 above)
Comment (vikramhh) [reply hack/make.ps1]: Thanks for catching the typo - fixed.

File: hack/make.ps1 (hunk 2 above)
Comment (vikramhh) [reply hack/make.ps1]: @thaJeztah - Would it be something like so:

1. Introduce a new Dockerfile for building gotestsum
2. From the container that we currently execute unit tests in, use that Dockerfile to build gotestsum in another container 
3. Copy the built gotestsum binary back to the unit tests container.

**Comment by andrewhsu** [comment]
Hmm...looks like the windows RS5 node was [not allocated properly](https://ci.docker.com/public/job/moby/job/PR-39971/4/execution/node/72/log/) to even start the tests:
```
 azwin-2-842900 was marked offline: Connection was broken: java.util.concurrent.TimeoutException: Ping started at 1569247809132 hasn't completed by 1569248049133
 	at hudson.remoting.PingThread.ping(PingThread.java:134)
 	at hudson.remoting.PingThread.run(PingThread.java:90)
```

I've [re-kicked](https://ci.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39971/5/pipeline) the job.

**Comment by andrewhsu** [comment]
You're going to need a section in `Jenkinsfile` under the windows stages to collect the junit xml files like so:
https://github.com/moby/moby/blob/9b07a3e0a59dbdabd1e3c71974a0e512f964d169/Jenkinsfile#L401-L405

After you add that, the additional tests will start showing up on the test report: https://ci.docker.com/public/job/moby/view/change-requests/job/PR-39971/test/?width=800&height=600

And under the test results page of the specific build number, e.g.: https://ci.docker.com/public/job/moby/view/change-requests/job/PR-39971/5/testReport/

**Comment by vikramhh** [comment]
@andrewhsu - thanks, pushed the Jenkinsfile change as well. RS5 for now, will add for RS1 as well after validation.

File: Jenkinsfile [hunk 3]
```diff
@@ -828,6 +828,7 @@ pipeline {
                     }
                     post {
                         always {
+                            junit testResults: 'bundles/junit-report.xml', allowEmptyResults: true
```
Comment (andrewhsu) [inline Jenkinsfile]: @vikramhh hmm...i don't see the test results for RS5: https://ci.docker.com/public/job/moby/job/PR-39971/6/testReport/

does windows care about the direction of the slashes in filename paths?

**Comment by andrewhsu** [comment]
Could you also have `Jenkinsfile` save the RS5 `junit-report.xml` to a zip file artifact https://ci.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39971/6/artifacts

It will help with diagnosis and processing of test results.

**Comment by thaJeztah** [comment]
I think I'll have to carry this one so that the Jenkinsfile changes are picked up

**Comment by thaJeztah** [comment]
opened https://github.com/moby/moby/pull/39998

**Comment by thaJeztah** [comment]
I want to try a different approach; perhaps Jenkins doesn't look at where the PR _came from_ (which fork), but only _who opened_ the PR. So instead of forking your branch to my GitHub fork, I want to open a PR from your branch.

GitHub doesn't allow having two PRs open from the same branch though, so I'll close this one, and open a new one from the same branch on your fork.

**Pull request closed without merging**